### PR TITLE
Fix PriceInfo usage with non-one quantities (SHOOP-1462)

### PR DESCRIPTION
**Pull request by suutari-ai**
Also set the correct discount amounts for SourceLine, BasketLine and
OrderLine when they are created.

Refs SHOOP-975 / SHOOP-1462

### CODE REVIEW

File: shoop/core/pricing/__init__.py [hunk 1]
```diff
@@ -161,7 +161,8 @@ def get_price_infos(self, context, products, quantity=1):
             for product_id in product_ids
         }
 
-    def get_pricing_steps_for_products(self, context, product_ids):
+    def get_pricing_steps_for_products(self, context, products):
```
Comment (tulimaki) [inline shoop/core/pricing/__init__.py]: Missing docstring.

Comment (suutari-ai) [inline shoop/core/pricing/__init__.py]: Yeah, was already missing. Feel free to add that. ;)

**Comment by tulimaki** [comment]
Now products get_price function has quantity 1 by default. Should we make sure that quantity is at least minimum purchase amount allowed? Would make it more simple to get from templates etc...

File: shoop/core/models/shops.py [hunk 2]
```diff
@@ -40,3 +41,15 @@ class Shop(TranslatableModel):
 
     def __str__(self):
         return self.safe_translation_getter("name", default="Shop %d" % self.pk)
+
+    def create_price(self, value):
+        """
+        Create a price with given value and settings of this shop.
+
+        Takes the ``prices_include_tax`` setting into account, and in
+        future might also do the same for a currency setting.
```
Comment (akx) [inline shoop/core/models/shops.py]: I assume `value` is just an unadorned decimal?

(`:type:` annotations would be nice.)

**Comment by suutari-ai** [comment]
> Now products get_price function has quantity 1 by default. Should we make sure that quantity is at least minimum purchase amount allowed? Would make it more simple to get from templates etc...

Are you proposing something like this?
```python
def get_price_info(self, context, quantity=None):
    from shoop.core.pricing import get_pricing_module
    module = get_pricing_module()
    pricing_context = module.get_context(context)
    if quantity is None:
        shop_product = self.get_shop_instance(pricing_context.shop)
        quantity = shop_product.minimum_purchase_quantity
    return module.get_price_info(pricing_context, product=self, quantity=quantity)
```

Seems somewhat logical, but how about the `Product.get_cheapest_child_price`, what should it do, since the minimum_purchase_quantity could be different for each child?

File: shoop/core/models/shops.py (hunk 2 above)
Comment (suutari-ai) [inline shoop/core/models/shops.py]: Right. Yep, I'll add the type annotations.

**Comment by tulimaki** [comment]
Yes something like that. I guess cheapest child price should return price that you can purchase some children cheapest. It think it should always force to minimum_purchase_quantity.

**Comment by suutari-ai** [comment]
> Yes something like that. I guess cheapest child price should return price that you can purchase some children cheapest. It think it should always force to minimum_purchase_quantity.

Force to minimum_purchase_quantity of which product? (the parent or one of the children?)

Also, quantity=1 as default is used in many other places too, e.g. in `PricingModule.get_price_info`, and `PricingModule.get_price_infos` and some others. Maybe changing the default from 1 to minimum_purchase_quantity should be another issue?

(As a side note: `PricingModule` methods should probably not have a default value for the quantity)

File: shoop/core/models/shops.py (hunk 2 above)
Comment (suutari-ai) [inline shoop/core/models/shops.py]: Added.

**Comment by tulimaki** [comment]
@suutari-ai I think we should talk about this in some meeting. And it should be different task since it effects so many places.

**Comment by tulimaki** [comment]
great work

**Pull request merged**